src/function/gds/rec_joins.cpp

```diff
@@ -116,58 +118,56 @@ void SPAlgorithm::bind(const expression_vector& params, Binder* binder,
-        RJOutputWriter* rjOutputWriter)
-        : mm{mm}, globalFT{globalFT}, rjOutputWriter{rjOutputWriter} {}
-
-    std::mutex mtx;
-    storage::MemoryManager* mm;
-    processor::FactorizedTable* globalFT;
-    RJOutputWriter* rjOutputWriter;
-};
-
-class RJOutputWriterVC : public VertexCompute {
-public:
-    explicit RJOutputWriterVC(RJOutputWriterVCSharedState* sharedState) : sharedState{sharedState} {
-        localFT = std::make_unique<processor::FactorizedTable>(sharedState->mm,
-            sharedState->globalFT->getTableSchema()->copy());
-        localRJOutputWriter = sharedState->rjOutputWriter->copy();
+    explicit RJVertexCompute(storage::MemoryManager* mm, processor::GDSCallSharedState* sharedState,
+        std::unique_ptr<RJOutputWriter> writer)
+        : mm{mm}, sharedState{sharedState}, writer{std::move(writer)} {
+        localFT = sharedState->claimLocalTable(mm);
     }
+    ~RJVertexCompute() override { sharedState->reclaimLocalTable(localFT); }
 
     void beginOnTable(table_id_t tableID) override {
-        localRJOutputWriter->beginWritingForDstNodesInTable(tableID);
+        writer->beginWritingForDstNodesInTable(tableID);
     }
 
     void vertexCompute(nodeID_t nodeID) override {
-        if (localRJOutputWriter->skipWriting(nodeID)) {
+        if (writer->skipWriting(nodeID)) {
             return;
         }
-        localRJOutputWriter->write(*localFT, nodeID);
+        writer->write(*localFT, nodeID);
     }
 
     void finalizeWorkerThread() override {
-        std::unique_lock lck(sharedState->mtx);
-        sharedState->globalFT->merge(*localFT);
+        // Do nothing.
     }
 
     std::unique_ptr<VertexCompute> copy() override {
-        return std::make_unique<RJOutputWriterVC>(sharedState);
+        return std::make_unique<RJVertexCompute>(mm, sharedState, writer->copy());
     }
 
 private:
-    RJOutputWriterVCSharedState* sharedState;
-    std::unique_ptr<processor::FactorizedTable> localFT;
-    std::unique_ptr<RJOutputWriter> localRJOutputWriter;
+    storage::MemoryManager* mm;
+    // Shared state storing ftables to materialize output.
+    processor::GDSCallSharedState* sharedState;
+    processor::FactorizedTable* localFT;
+    std::unique_ptr<RJOutputWriter> writer;
 };
 
 void RJAlgorithm::exec(processor::ExecutionContext* executionContext) {
+    auto clientContext = executionContext->clientContext;
     auto inputNodeMaskMap = sharedState->getInputNodeMaskMap();
     for (auto& tableID : sharedState->graph->getNodeTableIDs()) {
         if (!inputNodeMaskMap->containsTableID(tableID)) {
             continue;
         }
+        auto localTable = clientContext->getTx()->getLocalStorage()->getLocalTable(tableID);
+        if (localTable && localTable->cast<storage::LocalNodeTable>().getNumRows()) {
+            throw RuntimeException("Current recursive algorithm does not work with uncommited "
```

Comment:
Can you open an issue about this and assign to Guodong. Let's keep track of these things.